src/ol/tileurlfunction.js

```diff
@@ -26,17 +22,16 @@ export function createFromTemplate(template, tileGrid) {
       if (!tileCoord) {
         return undefined;
       }
-      return template
-        .replace(zRegEx, tileCoord[0].toString())
-        .replace(xRegEx, tileCoord[1].toString())
-        .replace(yRegEx, tileCoord[2].toString())
-        .replace(dashYRegEx, function () {
-          const z = tileCoord[0];
-          const range = tileGrid.getFullTileRange(z);
-          assert(range, 55); // The {-y} placeholder requires a tile grid with extent
-          const y = range.getHeight() - tileCoord[2] - 1;
-          return y.toString();
-        });
+      let maxY;
+      const z = tileCoord[0];
+      if (tileGrid) {
+        // TODO: fix places where we call this without a tile grid (tests fail without the above condition)
```

Comment:
Here are example test failures without this condition: https://github.com/openlayers/openlayers/actions/runs/3401521368/jobs/5656684976

With the previous code, there would be an unhandled error if the template included a `{-y}` placeholder - which is probably why we didn't notice.  Now the error happens at source construction.